### PR TITLE
MINIFICPP-2045 - Synchronous flow file restore and orphan content cleanup

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/rocksdb-repos/FlowFileRepository.h [hunk 1]
```diff
@@ -66,16 +66,14 @@ class FlowFileRepository : public ThreadedRepository, public SwapManager {
   }
 
   explicit FlowFileRepository(const std::string& repo_name = "",
-                     std::filesystem::path checkpoint_dir = FLOWFILE_CHECKPOINT_DIRECTORY,
+                     std::filesystem::path /*checkpoint_dir*/ = FLOWFILE_CHECKPOINT_DIRECTORY,
```
Comment (fgerlits) [inline extensions/rocksdb-repos/FlowFileRepository.h:69]: could this parameter be removed?  as far as I can see, this constructor is only used in a few unit tests (and from the other constructor, but that wouldn't need to be changed)

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp [hunk 2]
```diff
@@ -179,6 +180,37 @@ std::shared_ptr<io::BaseStream> DatabaseContentRepository::write(const minifi::R
...
 }
 
+bool DatabaseContentRepository::clearOrphans() {
+  if (!is_valid_ || !db_)
+    return false;
+  auto opendb = db_->open();
+  if (!opendb) {
+    return false;
+  }
+  std::vector<std::string> keys;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp]: renamed

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp [hunk 3]
```diff
@@ -179,6 +180,37 @@ std::shared_ptr<io::BaseStream> DatabaseContentRepository::write(const minifi::R
...
+  }
+  auto batch = opendb->createWriteBatch();
+  for (auto& key : keys) {
+    batch.Delete(key);
+  }
+
+  rocksdb::Status status = opendb->Write(rocksdb::WriteOptions(), &batch);
+
+  if (!status.ok()) {
+    logger_->log_debug("Could not delete orphan contents from rocksdb database: %s", status.ToString());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp]: changed it to error

File: extensions/rocksdb-repos/FlowFileRepository.cpp [hunk 4]
```diff
@@ -126,38 +123,28 @@ void FlowFileRepository::run() {
...
-    } else {
-      logger_->log_error("Couldn't open checkpoint database at '%s' using live database", checkpoint_dir_.string());
-      opendb = db_->open();
-    }
-    if (!opendb) {
-      logger_->log_trace("Could not open neither the checkpoint nor the live database.");
-      return;
+bool FlowFileRepository::ExecuteWithRetry(const std::function<rocksdb::Status()>& operation) {
+  std::chrono::milliseconds waitTime = 0ms;
+  for (int i=0; i < 3; ++i) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/rocksdb-repos/FlowFileRepository.cpp]: added constant

File: extensions/rocksdb-repos/FlowFileRepository.cpp [hunk 5]
```diff
@@ -126,38 +123,28 @@ void FlowFileRepository::run() {
...
-      logger_->log_trace("Successfully opened checkpoint database at '%s'", checkpoint_dir_.string());
-    } else {
-      logger_->log_error("Couldn't open checkpoint database at '%s' using live database", checkpoint_dir_.string());
-      opendb = db_->open();
-    }
-    if (!opendb) {
-      logger_->log_trace("Could not open neither the checkpoint nor the live database.");
-      return;
+bool FlowFileRepository::ExecuteWithRetry(const std::function<rocksdb::Status()>& operation) {
+  std::chrono::milliseconds waitTime = 0ms;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/rocksdb-repos/FlowFileRepository.cpp]: renamed

File: extensions/rocksdb-repos/FlowFileRepository.cpp [hunk 6]
```diff
@@ -191,62 +178,12 @@ void FlowFileRepository::prune_stored_flowfiles() {
...
-  if (!checkpoint_status.ok()) {
-    logger_->log_error("Could not create checkpoint object: %s", checkpoint_status.ToString());
-    return;
-  }
-  checkpoint_status = checkpoint->CreateCheckpoint(checkpoint_dir_.string());
-  if (!checkpoint_status.ok()) {
-    logger_->log_error("Could not initialize checkpoint: %s", checkpoint_status.ToString());
-    return;
+  flush();
+  if (content_repo_->clearOrphans()) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply extensions/rocksdb-repos/FlowFileRepository.cpp]: changed it to void, added some extra logs

File: extensions/rocksdb-repos/FlowFileRepository.h (hunk 1 above)
Comment (adamdebreceni) [reply extensions/rocksdb-repos/FlowFileRepository.h]: removed the parameter

**Review by martinzink — COMMENTED** [review]
It seems that rebasing broke the build. https://github.com/apache/nifi-minifi-cpp/commit/556407c86bfb641d1b64796105dc583e8f5dca25#diff-83f83358bb408f55a5cb82796cc807082ff8ebe1401d79b8949fadac41da76d8R271 uses the removed checkpoint_dir_ member.

Besides that LGTM

**Review by fgerlits — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**